waves_gateway/controller/flask_rest_controller.py

```diff
@@ -114,41 +155,68 @@ def _get_attempt_list(self) -> str:
         anything = flask_module.request.args.get('anything')  # type: Optional[str]
+        attempt = flask_module.request.args.get('attempt')  # type: Optional[int]
+        tries = flask_module.request.args.get('tries')  # type: Optional[int]
 
         trigger_receiver_as_int = None  # type: Optional[int]
 
         if trigger_receiver is not None:
             trigger_receiver_as_int = int(trigger_receiver)
 
-        if (trigger_tx is None) and (anything is None):
+        if (trigger_tx is None) and (anything is None) and (attempt is None) and (tries is None):
             return flask_module.abort(400)
 
         query = AttemptListQuery(
             trigger_currency=trigger_currency,
             trigger_tx=trigger_tx,
             trigger_receiver=trigger_receiver_as_int,
-            anything=anything)
+            anything=anything,
+            attempt=attempt,
+            tries=tries)
 
         attempt_lists = self._gateway_controller.query_attempt_lists(query)
 
         serialized_lists = map_array(lambda x: self._serializer.attempt_list_as_dict(x), attempt_lists)
 
         return flask_module.jsonify(serialized_lists)
 
-    def _check_coin_transaction(self, tx: str) -> str:
-        try:
-            self._gateway_controller.check_coin_transaction(tx)
-        except InvalidTransactionIdentifier:
-            flask_module.abort(400)
-        except NotImplementedError:
-            flask_module.abort(501)
+    def _check_coin_transaction(self, tx: str):
+        key = flask_module.request.headers['api-key']
+
+        if key == self._key:
+            try:
+                self._gateway_controller.check_coin_transaction(tx)
+            except InvalidTransactionIdentifier:
+                flask_module.abort(400)
+            except NotImplementedError:
+                flask_module.abort(501)
 
-        return ''
+            return ''
+        else:
+            flask_module.abort(403)
 
-    def _check_waves_transaction(self, tx: str) -> str:
-        try:
-            self._gateway_controller.check_waves_transaction(tx)
-        except InvalidTransactionIdentifier:
-            flask_module.abort(400)
+    def _get_block_heights(self):
+        heights = self._gateway_controller.get_block_heights()
+        print(heights)
+        return flask_module.jsonify(heights)
 
-        return ''
+    def _check_waves_transaction(self, tx: str):
+        key = flask_module.request.headers['api-key']
+
+        if key == self._key:
+            try:
+                self._gateway_controller.check_waves_transaction(tx)
+            except InvalidTransactionIdentifier:
+                flask_module.abort(400)
+
+            return ''
+        else:
+            flask_module.abort(403)
+
+    def _trigger_attemptlist_retry(self):
+        attempt_list_id = flask_module.request.args.get('attempt_list_id')  # type: Optional[str]
+        res = self._gateway_controller.trigger_attemptlist_retry(attempt_list_id)
+        if res:
```

Comment:
This is only allowed if res is a boolean which is not the case.

Reply:
fixed